src/libraries/System.Runtime/tests/System/IntPtrTests.cs

```diff
@@ -15,197 +17,215 @@ public static class IntPtrTests
+                Assert.Equal(expected, iEquatable.Equals(other));
             }
-            Assert.Equal(expected, ptr1.Equals(obj));
-            Assert.Equal(ptr1.GetHashCode(), ptr1.GetHashCode());
+            Assert.Equal(expected, value.Equals(obj));
+            Assert.Equal(value.GetHashCode(), value.GetHashCode());
         }
 
         [ConditionalFact(nameof(Is64Bit))]
-        public static unsafe void ImplicitCast()
+        public static unsafe void TestExplicitCast()
         {
-            var ptr = new IntPtr(42);
+            nint value = 42;
+
+            MethodInfo[] methods = typeof(nint).GetMethods();
+
+            MethodInfo opExplicitFromInt32 = typeof(nint).GetMethod("op_Explicit", new Type[] { typeof(int) });
+            MethodInfo opExplicitToInt32 = methods.Single((methodInfo) => (methodInfo.Name == "op_Explicit") && (methodInfo.ReturnType == typeof(int)));
+
+            int i = (int)opExplicitToInt32.Invoke(null, new object[] { value });
+            Assert.Equal(42, i);
+            Assert.Equal(value, (nint)opExplicitFromInt32.Invoke(null, new object[] { i }));
 
-            uint i = (uint)ptr;
-            Assert.Equal(42u, i);
-            Assert.Equal(ptr, (IntPtr)i);
+            MethodInfo opExplicitFromInt64 = typeof(nint).GetMethod("op_Explicit", new Type[] { typeof(long) });
+            MethodInfo opExplicitToInt64 = methods.Single((methodInfo) => (methodInfo.Name == "op_Explicit") && (methodInfo.ReturnType == typeof(long)));
 
-            ulong l = (ulong)ptr;
+            long l = (long)opExplicitToInt64.Invoke(null, new object[] { value });
             Assert.Equal(42u, l);
-            Assert.Equal(ptr, (IntPtr)l);
+            Assert.Equal(value, (nint)opExplicitFromInt64.Invoke(null, new object[] { l }));
 
-            void* v = (void*)ptr;
-            Assert.Equal(ptr, (IntPtr)v);
+            MethodInfo opExplicitFromPointer = typeof(nint).GetMethod("op_Explicit", new Type[] { typeof(void*) });
+            MethodInfo opExplicitToPointer = methods.Single((methodInfo) => (methodInfo.Name == "op_Explicit") && (methodInfo.ReturnType == typeof(void*)));
 
-            ptr = new IntPtr(0x7fffffffffffffff);
-            Assert.Throws<OverflowException>(() => (int)ptr);
+            void* v = Pointer.Unbox(opExplicitToPointer.Invoke(null, new object[] { value }));
+            Assert.Equal(value, (nint)opExplicitFromPointer.Invoke(null, new object[] { Pointer.Box(v, typeof(void*)) }));
+
+            value = unchecked((nint)0x7fffffffffffffff);
+            Exception ex = Assert.Throws<Exception>(() => opExplicitToInt32.Invoke(null, new object[] { value }));
+
+            if (ex is TargetInvocationException)
+            {
+                // RyuJIT throws TargetInvocationException wrapping an OverflowException
+                // while Mono directly throws the OverflowException
+                ex = ex.InnerException;
+            }
+            Assert.IsType<OverflowException>(ex);
```

Comment:
Turns out Mono and RyuJIT have different behavior for how the exception is thrown here.